### PR TITLE
Rewrite mod commands and add softban command

### PR DESCRIPTION
**Pull request by ilevn**

### CODE REVIEW

**Comment by SunDwarf** [comment]
In the pursuit of PEP 8 you made it unreadable

**Comment by ilevn** [comment]
At least my build doesnt fail ;)

File: dbansbot/cogs/mod.py [hunk 1]
```diff
@@ -45,51 +46,103 @@ def __init__(self, bot: DBans):
...
-                return
+        server = ctx.message.server
+
+        # Get the highest role
+        bot_highest = get_highest_role(server.me)
+
+        # Check the permission scope of the bot
+        if not channel.permissions_for(server.me).ban_members is True:
+            await self.bot.say(
+                ":x: My role (`{role}`) does not have permission to modify this user.".format(role=bot_highest))
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: This is fucking ugly, why is the last bracket on the same line

File: dbansbot/cogs/mod.py [hunk 2]
```diff
@@ -45,51 +46,103 @@ def __init__(self, bot: DBans):
...
+            return
+
+        assert isinstance(ctx.message.server, discord.Server)
+
+        if user_is_higher(user, server.me) or (user.id == ctx.message.server.owner_id):
+            # Too high, we can't touch them.
+            await self.bot.say(":x: My role (`{role}`) does not have permission to modify this user.".format(
+                role=bot_highest.name if bot_highest.name != "@everyone" else "everyone"))
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: Again here

File: dbansbot/cogs/mod.py (hunk 1 above)
Comment (ilevn) [inline dbansbot/cogs/mod.py]: what would you propose

File: dbansbot/cogs/mod.py [hunk 3]
```diff
@@ -45,51 +46,103 @@ def __init__(self, bot: DBans):
...
+            return
+
+        # Try to ban the user.
+        await self.bot.ban(user)
+        await self.bot.say(":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been banned"
+                           " from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else ""))
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: Bad ternary, use `banned from the server for ` and `.format(user, r=reason or "No reason specified.")`

File: dbansbot/cogs/mod.py (hunk 1 above)
Comment (gdude2002) [inline dbansbot/cogs/mod.py]: Putting it on the next line

File: dbansbot/cogs/mod.py [hunk 4]
```diff
@@ -45,51 +46,103 @@ def __init__(self, bot: DBans):
...
-                else:
-                    await self.bot.say(
-                        "Alright, {0.name}#{0.discriminator} has been kicked from the server!".format(user))
-            else:
-                await self.bot.say("I don't have enough permissions to kick them, their role is too high!")
-                return
+        server = ctx.message.server
+
+        # Get the highest role
+        bot_highest = get_highest_role(server.me)
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: `server.me.highest_role`

Stop copypasting old code especially if it's already existing

File: dbansbot/cogs/mod.py [hunk 5]
```diff
@@ -108,7 +108,8 @@ def __init__(self, bot: DBans):
         # Try to kick the user.
         await self.bot.kick(user)
         await self.bot.say(":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been kicked"
-                           " from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else ""))
+                           " from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else "")
+                           )
```
Comment (gdude2002) [inline dbansbot/cogs/mod.py]: Better, but it should be indented to the same level as line 110

File: dbansbot/cogs/mod.py [hunk 6]
```diff
@@ -142,7 +143,8 @@ def __init__(self, bot: DBans):
         await self.bot.ban(user)
         await self.bot.unban(user.server, user)
         await self.bot.say(":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been softbanned"
-                           " from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else ""))
+                           " from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else "")
+                           )
```
Comment (gdude2002) [inline dbansbot/cogs/mod.py]: Better, but it should be indented to the same level as line 145

File: dbansbot/cogs/mod.py (hunk 1 above)
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: Also, `:x:` is bad

`\N{CROSS MARK}`

The only places I use the replacement form is bad and I'm replacing them soon

File: dbansbot/cogs/mod.py (hunk 5 above)
Comment (gdude2002) [inline dbansbot/cogs/mod.py]: Actually scratch that, let me show you

File: dbansbot/cogs/mod.py [hunk 7]
```diff
@@ -45,51 +46,105 @@ def __init__(self, bot: DBans):
...
+
+        if user_is_higher(user, server.me) or (user.id == ctx.message.server.owner_id):
+            # Too high, we can't touch them.
+            await self.bot.say(":x: My role (`{role}`) does not have permission to modify this user.".format(
+                role=bot_highest.name if bot_highest.name != "@everyone" else "everyone"))
+            return
+
+        # Try to kick the user.
+        await self.bot.kick(user)
+        await self.bot.say(":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been kicked"
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: What the fuck

File: dbansbot/cogs/mod.py (hunk 5 above)
Comment (gdude2002) [inline dbansbot/cogs/mod.py]:
```python
await self.bot.say(
    ":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been kicked "
    "from the server{r}!".format(user, r=" for `{}`".format(reason) if reason else "")
)
```

Like this.

File: dbansbot/cogs/auto.py [hunk 8]
```diff
@@ -43,7 +43,8 @@ def __init__(self, bot: DBans):
                     body=body
                 )
                 if nstatus != 200:
-                    self.bot.logger.error("Unable to create new server `{}`: API returned `{}`".format(server.name, njs["msg"]))
+                    self.bot.logger.error(
```
Comment (SunDwarf) [inline dbansbot/cogs/auto.py]: Set your line length to not be 80

File: dbansbot/cogs/mod.py [hunk 9]
```diff
@@ -45,51 +45,114 @@ def __init__(self, bot: DBans):
...
+        if user.top_role.position > bot_highest.position or (user.id == ctx.message.server.owner_id):
+            # Too high, we can't touch them.
+            await self.bot.say("\N{CROSS MARK} My role (`{role}`) does not have permission to modify this user.".format(
+                role=bot_highest.name if bot_highest.name != "@everyone" else "everyone")
+            )
+            return
+
+        # Try to ban the user.
+        await self.bot.ban(user)
+        await self.bot.say(":heavy_check_mark: Alright, {0.name}#{0.discriminator} has been banned"
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: `:heavy_check_mark:`

File: dbansbot/cogs/auto.py (hunk 8 above)
Comment (ilevn) [inline dbansbot/cogs/auto.py]: whats your prefered length

File: dbansbot/cogs/mod.py [hunk 10]
```diff
@@ -45,51 +45,115 @@ def __init__(self, bot: DBans):
...
+                role=bot_highest if bot_highest != "@everyone" else "everyone"))
+            return
+
+        # Attempt to softban (ban and unban) the target
+        await self.bot.ban(user)
+        await self.bot.unban(user.server, user)
+        await self.bot.say("\N{HEAVY CHECK MARK} Alright, {0.name}#{0.discriminator} has been softbanned"
+                           " from the server: `{r}`".format(user, r=reason or "No reason specified.")
+
+                           )
```
Comment (SunDwarf) [inline dbansbot/cogs/mod.py]: stray bracket

**Comment by SunDwarf** [comment]
+1

**Pull request merged**